In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
xls = pd.ExcelFile('Excel_Test_(1).xlsx')

In [3]:
#xls.sheet_names

In [4]:
df1 = xls.parse("Exercise 2", skiprows = 2)

In [5]:
pd.set_option('display.max_columns', 200) 
df1 = df1.rename(columns={"Unnamed: 4": "Discount Amount"})
df1 = df1.rename(columns={"Unnamed: 12": "Discount Amount.1"})
df1 = df1.rename(columns={"Discount": "Discount %"})
df1 = df1.rename(columns={"Discount.1": "Discount %.1"})
df1 = df1.drop(df1.index[0])

# Table A Answer; 

## Assumption: "between" is inclusive, i.e., "<= 300 for the 10% discount" versus exclusive, "<300 for the 10% discount". In certain querying language like SQL this is true; but in general "between" is not inclusive. This assumption proves problematic for Table B, but assuming "between" is exclusive is not an option for the reason provided in the "Table B Answer" section.

## Solution: ask team in charge of discounts for clarification of discount rules.

In [6]:
def simple_discount(df1):
    if df1["Sales (Quantity)"] < 200:
        return 0
    if df1["Sales (Quantity)"] <= 300:
        return 0.1
    if df1["Sales (Quantity)"] <= 400:
        return 0.2
    else:
        return 0.3

In [7]:
df1["Revenue before discount"] = df1["Sales (Quantity)"]*df1["Sales Price"]

df1["Discount %"] = df1.apply(simple_discount, axis = 1)

df1["Discount Amount"] = df1["Sales (Quantity)"]*(df1["Sales Price"]*(df1["Discount %"]))

df1["Revenue after discount"] = df1["Sales (Quantity)"]*(df1["Sales Price"]*(1-df1["Discount %"]))

df1.loc[:,"Sales (Quantity)":"Revenue after discount"]

Sales (Quantity)  Sales Price  Revenue before discount  Discount %  \
1              342.0         15.0                   5130.0         0.2   
2              188.0         15.0                   2820.0         0.0   
3              338.0         15.0                   5070.0         0.2   
4              447.0         10.0                   4470.0         0.3   
5              339.0         11.0                   3729.0         0.2   
6              182.0         14.0                   2548.0         0.0   
7              128.0         13.0                   1664.0         0.0   
8              111.0         12.0                   1332.0         0.0   
9              153.0         14.0                   2142.0         0.0   
10             306.0         17.0                   5202.0         0.2   
11             123.0         18.0                   2214.0         0.0   
12             162.0         17.0                   2754.0         0.0   
13             316.0         13.0                   4108.0         0.2   
14             137.0         18.0                   2466.0         0.0   
15             471.0         17.0                   8007.0         0.3   
16             145.0         11.0                   1595.0         0.0   
17             101.0         13.0                   1313.0         0.0   
18             181.0         18.0                   3258.0         0.0   
19             356.0         15.0                   5340.0         0.2   
20             123.0         19.0                   2337.0         0.0   
21             409.0         16.0                   6544.0         0.3   
22             264.0         13.0                   3432.0         0.1   
23             105.0         13.0                   1365.0         0.0   
24             451.0         12.0                   5412.0         0.3   
25             443.0         10.0                   4430.0         0.3   
26             358.0         12.0                   4296.0         0.2   
27             318.0         13.0                   4134.0         0.2   
28             414.0         16.0                   6624.0         0.3   
29             153.0         17.0                   2601.0         0.0   
30             301.0         12.0                   3612.0         0.2   
31             124.0         15.0                   1860.0         0.0   
32             406.0         10.0                   4060.0         0.3   
33             444.0         19.0                   8436.0         0.3   
34             150.0         14.0                   2100.0         0.0   
35             141.0         17.0                   2397.0         0.0   
36             147.0         15.0                   2205.0         0.0   
37             345.0         17.0                   5865.0         0.2   
38             182.0         12.0                   2184.0         0.0   
39             487.0         12.0                   5844.0         0.3   
40             392.0         12.0                   4704.0         0.2   
41             170.0         14.0                   2380.0         0.0   
42             372.0         13.0                   4836.0         0.2   
43             316.0         15.0                   4740.0         0.2   
44             241.0         16.0                   3856.0         0.1   
45             137.0         17.0                   2329.0         0.0   
46             246.0         16.0                   3936.0         0.1   
47             395.0         17.0                   6715.0         0.2   
48             342.0         12.0                   4104.0         0.2   
49             187.0         17.0                   3179.0         0.0   

    Discount Amount  Revenue after discount  
1            1026.0                  4104.0  
2               0.0                  2820.0  
3            1014.0                  4056.0  
4            1341.0                  3129.0  
5             745.8                  2983.2  
6               0.0        

# Table B Answer; 

### "Between" being inclusive Issue:

### Assuming between is inclusive, imagine a customer buys exactly 300 units and those units are exactly 16 dollars. Which discount should be applied? From a customer's perspective, the 20% discount. But if the rules are programmed in the order listed on the instructions for Table B, they will receive a 10% discount as the criteria "quantity between 200 and 300 and price between 12 - 16 dollars" (10% discount) criteria apply first. If we assume "Between" is not inclusive, a new problem emerges: When the customer buys exactly 300 units and those units are exactly 16 dollars, the computer will throw an error or will not give a discount as those values are no longer part of the formula, i.e.,  only < 16  or  > 16 are in the formula while =16 is not.

### That is, depending on the order and interpretation of the criteria provided, the results will differ. For example, I assume that customer 15 (units bought: 471, sale price: 17) should receive a 30% discount given the rules. However if the criteria are programmed in the order provided, the computer will give this person a 20% because they hit the "or price between 16 - 18 dollars" (20% discount) criteria before they hit the "quantity more than 400" (30% discount) criteria.

### Solution: ask team in charge of discounts for clarification of discount rules. For Python, I have gone with giving bigger discounts as happy customers are better for long-term growth than angry customers who feel they have been tricked by ambiguous language. That is, the solution my formula provides here and my answers given in my Excel solution differ.

In [8]:
def complex_discount(df1):
    if (df1["Sales (Quantity)"] > 400) or (df1["Sales Price"] > 18):
        return 0.3

    if (300 <= df1["Sales (Quantity)"] <= 400) or (16 <= df1["Sales Price"] <= 18):
        return 0.2
            
    if (200 <= df1["Sales (Quantity)"] <= 300) and (12 <= df1["Sales Price"] <= 16):
        return 0.1

    if (df1["Sales (Quantity)"] < 200) or (df1["Sales Price"] < 12):
        return 0            

In [9]:
df1["Revenue before discount.1"] = df1["Sales (Quantity).1"]*df1["Sales Price.1"]

df1["Discount %.1"] = df1.apply(complex_discount, axis = 1)

df1["Discount Amount.1"] = df1["Sales (Quantity).1"]*(df1["Sales Price.1"]*(df1["Discount %.1"]))

df1["Revenue after discount.1"] = df1["Sales (Quantity).1"]*(df1["Sales Price.1"]*(1-df1["Discount %.1"]))

df1.loc[:,"Sales (Quantity).1":"Revenue after discount.1"]

Sales (Quantity).1  Sales Price.1  Revenue before discount.1  \
1                342.0           15.0                     5130.0   
2                188.0           15.0                     2820.0   
3                338.0           15.0                     5070.0   
4                447.0           10.0                     4470.0   
5                339.0           11.0                     3729.0   
6                182.0           14.0                     2548.0   
7                128.0           13.0                     1664.0   
8                111.0           12.0                     1332.0   
9                153.0           14.0                     2142.0   
10               306.0           17.0                     5202.0   
11               123.0           18.0                     2214.0   
12               162.0           17.0                     2754.0   
13               316.0           13.0                     4108.0   
14               137.0           18.0                     2466.0   
15               471.0           17.0                     8007.0   
16               145.0           11.0                     1595.0   
17               101.0           13.0                     1313.0   
18               181.0           18.0                     3258.0   
19               356.0           15.0                     5340.0   
20               123.0           19.0                     2337.0   
21               409.0           16.0                     6544.0   
22               264.0           13.0                     3432.0   
23               105.0           13.0                     1365.0   
24               451.0           12.0                     5412.0   
25               443.0           10.0                     4430.0   
26               358.0           12.0                     4296.0   
27               318.0           13.0                     4134.0   
28               414.0           16.0                     6624.0   
29               153.0           17.0                     2601.0   
30               301.0           12.0                     3612.0   
31               124.0           15.0                     1860.0   
32               406.0           10.0                     4060.0   
33               444.0           19.0                     8436.0   
34               150.0           14.0                     2100.0   
35               141.0           17.0                     2397.0   
36               147.0           15.0                     2205.0   
37               345.0           17.0                     5865.0   
38               182.0           12.0                     2184.0   
39               487.0           12.0                     5844.0   
40               392.0           12.0                     4704.0   
41               170.0           14.0                     2380.0   
42               372.0           13.0                     4836.0   
43               316.0           15.0                     4740.0   
44               241.0           16.0                     3856.0   
45               137.0           17.0                     2329.0   
46               246.0           16.0                     3936.0   
47               395.0           17.0                     6715.0   
48               342.0           12.0                     4104.0   
49               187.0           17.0                     3179.0   

    Discount %.1  Discount Amount.1  Revenue after discount.1  
1            0.2             1026.0                    4104.0  
2            0.0                0.0                    2820.0  
3            0.2             1014.0                    4056.0  
4            0.3             1341.0                    3129.0  
5            0.2              745.8                    2983.2  
6            0.0                0.0                    2548.0  
7            0.0                0.0                    1664.0  
8            0.0                0.0                    1332.0  
9            0.0           